File: python/trab3/secantes.py
from base import *


@algoritmo
def secantes(self, f, xi, xf):
    self.tabela('k', 'xi', 'xf', 'x', 'f(xi)', 'f(xf)')
    for k in xrange(self.nmi):
        fxi, fxf = f(xi), f(xf)
        p = fxf / fxi
        x = xf - (xi - xf) * p / (1 - p)
        self.tabela(k, xi, xf, x, fxi, fxf)
        if abs(x - xf) < self.tol and abs(f(x)) < self.tol:
            return x
        xi, xf = xf, x
    raise ErroNMI

File: ruby/trab3/secantes.rb
# encoding: utf-8
# Algoritmo SECANTES
require "./base"

Calcnum::algoritmo :secantes do |x0, x1|
  tabela 'k', 'x0', 'x1', 'p', 'x'
  iterar do |k|
    p = f(x1) / f(x0)
    x = x1 - (x0 - x1) * p / (1.0 - p)
    tabela k, x0, x1, p, x
    break x if (x - x1).abs < tol and f(x).abs < tol
    x0, x1 = x1, x
  end
end

